File: baekjoon/cpp/solutions/9466.cpp
#include <iostream>

using namespace std;

int arr[100001];
int visited[100001]; // 0: 방문X, 1: 팀 생성 가능, -1: 팀 생성 불가능
int n, start;

void dfs(int i) {
	// 순환: x에서 시작, y가 순환의 시작 => x부터 y 전까지 visited를 -1로 변경
	// 단 x와 y와 같은 경우 pass
	// 순환 발생 여부: 접근하려고 하는 사람이 이미 방문한(y) 경우

	if (visited[i]) {
		if (i == start)
			return;
		while (start != i) {
			visited[start] = -1;
			start = arr[start];
		}
		return;
	}

	visited[i] = 1;
	dfs(arr[i]);
}

int main() {
	cin.tie(0);
	ios::sync_with_stdio(0);

	int T;
	cin >> T;
	while (T--) {
		cin >> n;
		for (int i = 1;i <= n;i++) {
			cin >> arr[i];
		}

		for (int i = 1;i <= n;i++) {
			start = i;
			if (visited[i] == 0) {
				dfs(i);
			}
		}

		int cnt = 0;
		for (int i = 1;i <= n;i++) {
			if (visited[i] != 1)
				cnt++;
			visited[i] = 0;
		}

		cout << cnt << "\n";
	}
}

File: baekjoon/python/9466.py
import sys
input=sys.stdin.readline
sys.setrecursionlimit(10**6)

def dfs(cur,start,arr,visited):
    # 이미 방문한 경우
    if visited[cur]:
        # 본인이 본인을 선택한 경우
        if cur==start:
            return
        
        # start부터 cur 이전까지는 팀 생성 불가능
        while start!=cur:
            visited[start]=-1
            start=arr[start]

        return
    
    visited[cur]=1
    dfs(arr[cur],start,arr,visited)

T = int(input())

for _ in range(T):
    n=int(input())
    arr = [0] + list(map(int, input().split()))

    # 0: 방문 X, 1: 팀 생성 가능, -1: 팀 생성 불가능
    visited=[0]*(n+1)

    for i in range(1, n+1):
        if visited[i]==0:
            start=i
            dfs(i,start,arr,visited)

    cnt=0
    for i in range(1, n+1):
        if visited[i]!=1:
            cnt+=1

    print(cnt)